model/riscv_insts_vext_vset.sail

```diff
@@ -71,60 +71,54 @@ mapping maybe_ma_flag : string <-> bits(1) = {
-union clause ast = VSET_TYPE : (vsetop, bits(1), bits(1), bits(3), bits(3), regidx, regidx)
+val handle_illegal_vtype : unit -> unit effect {rreg}
+function handle_illegal_vtype() = {
+  /* Note: Implementations can set vill or trap if the vtype setting is not supported.
+   * TODO: configuration support for both solutions
+   */
+  vtype->bits() = 0b1 @ zeros(sizeof(xlen) - 1); /* set vtype.vill */
+  vl = zeros();
+  print_reg("CSR vtype <- " ^ BitStr(vtype.bits()));
+  print_reg("CSR vl <- " ^ BitStr(vl))
+}
 
-mapping encdec_vsetop : vsetop <-> bits(4) ={
-  VSETVLI <-> 0b0000,
-  VSETVL  <-> 0b1000
+val calculate_new_vl : (int, int) -> xlenbits
+function calculate_new_vl(AVL, VLMAX) = {
+  /* Note: ceil(AVL / 2) ≤ vl ≤ VLMAX when VLMAX < AVL < (2 * VLMAX)
+   * TODO: configuration support for either using ceil(AVL / 2) or VLMAX
+   */
+  if AVL <= VLMAX then to_bits(sizeof(xlen), AVL)
+  else if AVL < 2 * VLMAX then to_bits(sizeof(xlen), (AVL + 1) / 2)
+  else to_bits(sizeof(xlen), VLMAX)
 }
 
-mapping clause encdec = VSET_TYPE(op, ma, ta, sew, lmul, rs1, rd) if haveVExt()
-  <-> encdec_vsetop(op) @ ma @ ta @ sew @ lmul @ rs1 @ 0b111 @ rd @ 0b1010111 if haveVExt()
+/* *********************************** vsetvli *********************************** */
+union clause ast = VSETVLI : (bits(1), bits(1), bits(3), bits(3), regidx, regidx)
 
-function clause execute VSET_TYPE(op, ma, ta, sew, lmul, rs1, rd) = {
-  let VLEN_pow      = get_vlen_pow();
-  let ELEN_pow      = get_elen_pow();
+mapping clause encdec = VSETVLI(ma, ta, sew, lmul, rs1, rd) if haveVExt()
+  <-> 0b0000 @ ma @ ta @ sew @ lmul @ rs1 @ 0b111 @ rd @ 0b1010111 if haveVExt()
+
+function clause execute VSETVLI(ma, ta, sew, lmul, rs1, rd) = {
   let LMUL_pow_ori  = get_lmul_pow();
   let SEW_pow_ori   = get_sew_pow();
   let ratio_pow_ori = SEW_pow_ori - LMUL_pow_ori;
 
   /* set vtype */
-  match op {
-    VSETVLI => {
-      vtype->bits() = 0b0 @ zeros(sizeof(xlen) - 9) @ ma @ ta @ sew @ lmul
-    },
-    VSETVL  => {
-      let rs2 : regidx = sew[1 .. 0] @ lmul;
-      vtype->bits() = X(rs2)
-    }
-  };
+  vtype->bits() = 0b0 @ zeros(sizeof(xlen) - 9) @ ma @ ta @ sew @ lmul;
 
-  /* check legal SEW and LMUL and calculate VLMAX */
+  /* check new SEW and LMUL are legal and calculate VLMAX */
+  let VLEN_pow     = get_vlen_pow();
+  let ELEN_pow     = get_elen_pow();
   let LMUL_pow_new = get_lmul_pow();
   let SEW_pow_new  = get_sew_pow();
-  if SEW_pow_new > LMUL_pow_new + ELEN_pow then {
-    /* Note: Implementations can set vill or trap if the vtype setting is not supported.
-     * TODO: configuration support for both solutions
-     */
-    vtype->bits() = 0b1 @ zeros(sizeof(xlen) - 1); /* set vtype.vill */
-    vl = zeros();
-    print_reg("CSR vtype <- " ^ BitStr(vtype.bits()));
-    print_reg("CSR vl <- " ^ BitStr(vl));
-    return RETIRE_SUCCESS
-  };
+  if SEW_pow_new > LMUL_pow_new + ELEN_pow then { handle_illegal_vtype(); return RETIRE_SUCCESS };
```

Comment:
I know that given operator precende rules this work but I would prefer parenthesis for clarity to make clear what is compared with.

Reply:
Parentheses are added to related lines.